ros1_foxglove_bridge/src/ros1_foxglove_bridge_nodelet.cpp

```diff
@@ -32,24 +33,29 @@ class FoxgloveBridge : public nodelet::Nodelet {
   FoxgloveBridge() = default;
   virtual void onInit() {
     auto& nhp = getPrivateNodeHandle();
+    const auto host = nhp.param<std::string>("host", DEFAULT_HOST);
     const int port = nhp.param<int>("port", DEFAULT_PORT);
     const int max_update_ms = nhp.param<int>("max_update_ms", DEFAULT_MAX_UPDATE_MS);
 
     ROS_INFO("Starting %s with %s", ros::this_node::getName().c_str(),
              foxglove::WebSocketUserAgent());
 
-    _server = std::make_unique<foxglove::Server<foxglove::WebSocketNoTls>>(
-      "foxglove_bridge",
-      std::bind(&FoxgloveBridge::logHandler, this, std::placeholders::_1, std::placeholders::_2));
-    _server->setSubscribeHandler(
-      std::bind(&FoxgloveBridge::subscribeHandler, this, std::placeholders::_1));
-    _server->setUnsubscribeHandler(
-      std::bind(&FoxgloveBridge::unsubscribeHandler, this, std::placeholders::_1));
-    _server->start(static_cast<uint16_t>(port));
-
-    _msgParser = std::make_unique<foxglove_bridge::MsgParser>();
-    _updateTimer = getMTNodeHandle().createTimer(ros::Duration(max_update_ms / 1e3),
-                                                 &FoxgloveBridge::updateAdvertisedTopics, this);
+    try {
+      _server = std::make_unique<foxglove::Server<foxglove::WebSocketNoTls>>(
+        "foxglove_bridge",
+        std::bind(&FoxgloveBridge::logHandler, this, std::placeholders::_1, std::placeholders::_2));
+      _server->setSubscribeHandler(
+        std::bind(&FoxgloveBridge::subscribeHandler, this, std::placeholders::_1));
+      _server->setUnsubscribeHandler(
+        std::bind(&FoxgloveBridge::unsubscribeHandler, this, std::placeholders::_1));
+      _server->start(host, static_cast<uint16_t>(port));
+
+      _msgParser = std::make_unique<foxglove_bridge::MsgParser>();
+      _updateTimer = getMTNodeHandle().createTimer(ros::Duration(max_update_ms / 1e3),
+                                                   &FoxgloveBridge::updateAdvertisedTopics, this);
+    } catch (const std::exception& err) {
+      ROS_ERROR("Failed to start websocket server: %s", err.what());
```

Comment:
This is not a good state to leave the node in. There is no retry loop or ability to try again, so if the server failed to start it would print to a log somewhere and just be a zombie node at that point. Better to crash the node.

Reply:
It's a bit more involved as the server is loaded as a nodelet. So if you would try to exit the process then the nodelet manager and all other loaded nodelets would terminate as well. Will check what the common way to handle this is.

Reply:
I noticed rosbridge has some somewhat complicated retry loop code. Maybe we can come up with something simpler, but that's always a last resort option.

Reply:
Nodelets that fail to initialize are unloaded again:

https://github.com/ros/nodelet_core/blob/6265cac6bd686a7feb872bba5304f16de4f0fd46/nodelet/src/loader.cpp#L314-L326

The referenced exception handler doesn't log the specific error message, so I would catch it to log an error and then retrow it.